docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+# Secure Impersonation
+
+* [Context](#context)
+* [Goals](#goals)
+* [Design](#design)
+  - [controller serviceAccounts](#controller-serviceaccounts)
+  - [user/group names](#user-group-names)
+  - [serviceAccount impersonation](#serviceaccount-impersonation)
+  - [defaulting](#defaulting)
+  - [kubeConfig](#kubeconfig)
+  - [sourceRefs](#sourcerefs)
+  - [kubectl](#kubectl)
+* [Concerns](#concerns)
+  - [compatibiliity](#compatibiliity)
+  - [performance](#performance)
+* [Decisions](#decisions)
+* [Tasks](#tasks)
+* [Examples](#examples)
+  - [default install](#default-install)
+  - [tenants](#tenants)
+  - [cross-namespace binding against Flux Groups](#cross-namespace-binding-against-flux-groups)
+  - [opening Source policies with RBAC](#opening-source-policies-with-rbac)
+  - [rejecting kubeConfigs](#rejecting-kubeconfigs)
+
+## Context
+
+Flux 2's installation model is built from CRD's.
+This allows cluster operators to granularly control who can create/modify Sources and Applier type objects such as
+GitRepository and Kustomization and for which Namespaces.
+This also helps out the Deployment model.
+While we could for example, still run independent instances of ex:kustomize-controller in each Namespace listening to only
+Kustomize resources for that Namespace, it can use fewer resources to deploy a single controller watching all (or many) Namespaces.
+Having a single controller reduces the onboarding steps necessary for a new Namespace owner to use Flux Custom Resources.
+A central set of controllers is the primary Deployment mode for Flux.
+
+In order to manage resources across many Namespaces, the default install gives kustomize-controller and helm-controller a full
+cluster-admin ClusterRoleBinding.
+This means cluster owners need to be careful who can create Kustomizations and HelmReleases as well as which Sources are in the cluster
+due to the exposure they provide to act as cluster-admin.
+This restricts who can create Source and Apply type objects into a burdensome approval flow for cluster owners and provides
+poor delegation of self-service permissions which is a proven elements of good platforms.
+
+
+## Goals
+
+Flux should remain easy to install but promote secure practices in its default configuration.
+Flux's security model should allow a cluster administrator to delegate self-service management of
+Sources, Applies, Webhooks, Automations, and Notifications to Namespace owners/users without having to review every detail for
+compliance to security policies.
+Tenants or Namespace owners should be able to act freely without fear that their changes will overreach their permissions.
+
+Flux APIs must structurally prevent privilege escalation and arbitrary code execution.
+Flux APIs should be congruent with Kubernetes core API's security model.
```

Comment:
** Assuming the proposal is geared towards the Namespace-as-a-Service [tenancy model](https://kubernetes.io/blog/2021/04/15/three-tenancy-models-for-kubernetes/). **

The challenge here from a security perspective is providing _privilege escalation and arbitrary code execution_ prevention guarantees to flux's users. 
The proposed changes focuses in providing isolation and least privileges at control plane level only.
Privesc routes from Pod to Host, would not be contained. In the same way as lateral movement through to other tenants would be open for abuse by a malicious tenant.  

At the same time, trying to enforce those levels of isolation on behalf of flux's users may become disruptive and break compatibility with existing workloads. That would entail becoming more opinionated around some security primitives (i.e. Pod Security Standards, workload scheduling, etc), which feels beyond flux's scope.
So I am not sure it is worth pursuing _privilege escalation and arbitrary code execution_ prevention as a goal.

In my opinion the proposal could highlight as a goal making flux _optimised_ for multi-tenancy, focusing on a least privileged security model at *control plane* and better resource utilisation instead.

Reply:
Good point @pjbgf, I think we should limit the scope to “resources reconciliation” and be explicit about what this means. The model proposed here is about giving cluster admins a way to define polices that restrict what a tenant source (Git, Helm, Bucket, etc) can contain and thus limiting the operations that tenants can perform on a cluster. Flux can ensure that a tenant can deploy applications in certain namespaces and only in those. Flux can’t ensure that the applications deployed are themselves secure, this is in scope for network policy, PsP, seccomp profiles, etc.